nixos/modules/services/misc/matrix-conduit.nix

```diff
@@ -0,0 +1,233 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.matrix-conduit;
+
+  format = pkgs.formats.toml {};
+  configFile = format.generate "conduit.toml" cfg.settings;
+in
+  {
+    meta.maintainers = with maintainers; [ pstn piegames ];
+    options.services.matrix-conduit = {
+      enable = mkEnableOption "matrix-conduit";
+      iHaveReadTheFederationWarning = mkOption {
+        type = types.bool;
+        default = false;
+        example = true;
+        description = ''
+          Conduit upstream documentation contains the following warning:
+          You should not join Matrix rooms without asking the admins first.
+          We do not know whether Conduit is safe for general use yet, so you
+          should assume there is some chance that it breaks rooms permanently
+          for all participating users. We are not aware of such a bug today,
+          but we would like to do more testing.
+
+          This option will be removed upon removal of this warning from the
+          upstream documentation.
+        '';
+      };
+
+
+      extraEnvironment = mkOption {
+        type = types.attrsOf types.str;
+        description = "Extra Environment variables to pass to the conduit server.";
+        default = {};
+        example = { RUST_BACKTRACE="yes"; };
+      };
+
+      nginx.enable = mkOption {
+        type = types.bool;
+        default = false;
+        example = true;
+        description = ''
+          Whether to enable a nginx vhost that will listen on all interfaces on tcp/443
+          for https connections and proxy them to conduit. Further nginx configuration
+          can be done by adapting <option>services.nginx.virtualHosts.&lt;settings.global.server_name&gt;</option>.
+          When this is enabled, ACME will be used to retrieve a TLS certificate by default. To disable
+          this, set the <option>services.nginx.virtualHosts.&lt;settings.global.server_name&gt;.enableACME</option> to
+          <literal>false</literal> and if appropriate do the same for
+          <option>services.nginx.virtualHosts.&lt;settings.global.server_name&gt;.forceSSL</option>.
+        '';
+      };
+      package = mkOption {
+        type = types.package;
+        default = pkgs.matrix-conduit;
+        defaultText = "pkgs.matrix-conduit";
+        example = "pkgs.matrix-conduit";
+        description = ''
+          Package of the conduit matrix server to use.
+        '';
+      };
+
+      settings = mkOption {
+        type = types.submodule {
+          freeformType = format.type;
+          options = {
+            global.server_name = mkOption {
+              type = types.nullOr types.str;
```

Comment:
Why allow `null`? https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml#L11-12

You could then also get rid of the not null assertion further down.

Reply:
It's null because there is no sensible default for the field. `localhost` could be one but I doubt many people want to chat on localhost via matrix.

Reply:
Fields don't need to have a default value. In fact, not having a default is the whole point of required fields (no default <=> required).

Reply:
So should I just not assert anything and let nix fail normally when `server_name` isn't set?

Reply:
That's how it's usually done, yes.